cpp/src/jit/parser.cpp

```diff
@@ -30,36 +31,24 @@ inline bool is_white(const char c) { return c == ' ' || c == '\n' || c == '\r' |
 
 std::string ptx_parser::escape_percent(const std::string& src)
 {
-  // Since we are transforming into inline ptx we are not allowed to have
-  // register names starting with %.
-  const size_t length = src.size();
-  size_t start        = 0;
-  size_t stop         = 0;
-  while (start < length && (is_white(src[start]) || src[start] == '[')) { start++; }
-  stop = start;
-  while (stop < length && !is_white(src[stop]) && src[stop] != ']') { stop++; }
-  if (src[start] == '%') {
+  // b/c we're transforming into inline ptx we aren't allowed to have register names starting with %
+  auto f = std::find_if_not(src.begin(), src.end(), [](auto c) { return is_white(c) || c == '['; });
+  if (f != src.end() && *f == '%') {
     std::string output = src;
-    output.replace(start, 1, percent_escape);
+    output.replace(std::distance(src.begin(), f), 1, percent_escape);
     return output;
-  } else {
-    return src;
   }
+  return src;
 }
 
 std::string ptx_parser::remove_nonalphanumeric(const std::string& src)
 {
-  const size_t length = src.size();
-  size_t start        = 0;
-  size_t stop         = 0;
-  std::string output  = src;
-  while (start < length && (is_white(src[start]) || src[start] == '[')) { start++; }
-  stop = start;
-  while (stop < length && !is_white(src[stop]) && src[stop] != ']') {
-    if (!isalnum(src[stop]) && src[stop] != '_') { output[stop] = '_'; }
-    stop++;
-  }
-  return output.substr(start, stop - start);
+  std::string out = src;
+  auto f = std::find_if_not(out.begin(), out.end(), [](auto c) { return is_white(c) || c == '['; });
+  auto l = std::find_if(out.begin(), out.end(), [](auto c) { return is_white(c) || c == ']'; });
```

Comment:
This still doesn't look right to me. Shouldn't the `find_if` start from where the `find_if_not` left off in order to match the original code? It seems like this could incorrectly end up with `l` being before `f`.

Reply:
If the input PTX code is correct `]` should never appear before `[`. This parser is not intended to be a compiler and should not be designed to detect code errors, otherwise the logic of this whole file probably needs to be changed.

So I think both versions are fine but @trevorsm7's is slightly better, in terms of sanity.